Это вторая часть задания.

Первая часть задания (задания 3.1-3.4) находится в [ноутбуке про линейную регрессию](https://colab.research.google.com/drive/1eX21oBHRAzNr3PimxhMJENF3yeBC2DBe?usp=sharing).

# Предсказание цен на недвижимость

Компании по продаже недвижимости оценивают её стоимость, используя методы машинного обучения.
Задача оценки цены на недвижимость также может быть нужна для выявления наиболее выгодных предложений, или, например, для выявления аномально низких цен на объект недвижимости.

**Постановка задачи анализа данных**

Целью данной задачи является прогнозирование стоимости домов в округе Кинг (штат Вашингтон, США) с помощью построения регрессионных моделей и их анализа. Набор данных состоит из цен на дома в округе Кинг, проданных в период с мая 2014 года по май 2015 года. Данные опубликованы в открытом доступе на платформе Kaggle.

Данные с сайта renthub.com по стоимости квартир для аренды:

<a href="https://drive.google.com/uc?id=1Dr-p-o5yMi3uIS2kPcUUwT5bH0cCt1Be
" target="_blank"><img src="https://drive.google.com/uc?id=1Dr-p-o5yMi3uIS2kPcUUwT5bH0cCt1Be"
alt="IMAGE ALT TEXT HERE" width="360" border="0" /></a>


**Обзор доступных данных**

В наборе данных 21613 наблюдений и 16 признаков.
Таким образом, про каждый из 21613 объектов недвижимости мы знаем значения 16 характеристик (число спален, оценка состояния риелтором, наличие вида на воду и т.п.)

Выборка была разбита на две части для обучения и для тестирования модели. Данные на обучение и тест обычно делят случайно и равномерно: нужно чтобы и обучающая, и тестовая выборка были похожие на те данные, на которых модель будет использоваться в боевых условиях.

Более сложные разбиения могут использоваться если в данных есть временная компонента - например, мы строим скоринговую модель и для теста откладываем данные за последние несколько месяцев, а всю остальную выборку используем для обучения. Такое разбиение имитирует реальную работу модели: мы можем обучаться только на данных прошлых кредитов, а работать модель уже будет на данных из будущего.

Доли общей выборки для обучения и тестирования обычно 70% и 30% соотвественно. Любые разумные числа подходят, если для обучения используется достаточно много данных (обычно больше 50%), но и для тестирования что-то остается (10% и больше).


Данные содержат два типа переменных:

* Целевая: **Целевая. Цена**
* Остальные переменные **могут использоваться для прогноза целевой переменной.**

## План анализа данных (data mining):

  1. Загрузить данные для обучения
  2. Обработать данные перед обучением модели
  3. Обучить модель на обучающей выборке
  4. Загрузить и предобработать данные для тестирования
  5. Провалидировать модель на тестовой выборке
  6. Проанализировать важность признаков и не забыть про постановку задачи

## 1. Загрузить данные для обучения

**Шаг 1.1. Загружаем библиотеки**

Библиотека **warnings** отвечает за то, какие предупреждения (warnings) о работе будут выводиться пользователю.
FutureWarning - предупреждения о том, как изменится работа библиотек в будущих версиях.
Поэтому такие предупреждения мы будем игнорировать.
Чтобы включить режим игнорирования мы отбираем все предупреждения из категории FutureWarning и выбираем для них действия 'ignore'.
Это делается вызовом функции simplefilter c задание двух атрибутов: действия action и категории предупреждений category.

In [1]:
import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

Для корректной работы с данными в python требуется загрузить специальную библиотеку
**pandas**, программную библиотеку на языке python для обработки и анализа данных.

In [2]:
import pandas as pd # загружаем библиотеку и для простоты обращения в коде называем её сокращенно pd

Для корректной работы с графиками в python требуется загрузить специальную библиотеку
**matplotlib**, программную библиотеку на языке python для визуализации данных двумерной и трехмерной графикой.

Графики используются для облегчения интерпретации полученных результатов, а также в качестве иллюстраций в презентациях и отчетах.

Оснвные методы для построения:
* plot() - графики
* semilogy() - график логарифмический
* hist() - гистограммы

In [3]:
import matplotlib.pyplot as plt # загружаем библиотеку и для простоты обращения в коде называем её сокращенно plt
# указываем, чтобы картинки отображались прямо в ноутбуке
%matplotlib inline

**Шаг 1.2. Загрузим данные**

Для решения задачи мы будем использовать данные. Они состоят из двух частей: часть для обучения и часть для тестирования модели. Загружаем данные с помощие команды !wget. Для того, чтобы игнорировать сообщения в процессе загрузки используем магическую команду %%capture в первой строке.

In [4]:
%%capture
!wget https://www.dropbox.com/s/afwb0tnqm9izxha/predict_house_price_training_data.xlsx
!wget https://www.dropbox.com/s/sur2avqf4n5f4az/predict_house_price_test_data.xlsx

Так как данные в формате xlsx (Excel), мы будем использовать специальную функцию
из библиотеки pandas для загрузки таких данных **read_excel**.

В функции передаем один атрибут: название таблицы с данными.

In [5]:
training_data = pd.read_excel('predict_house_price_training_data.xlsx') # загружаем таблицу в переменную training_data

*Что важно посмотреть после того, как мы загрузили данные?*
- проверить, что данные действительно загрузились
- посмотреть на данные, чтобы удостовериться, что они правильные: колонки имеют те же названия, что и в таблице и т.д.

Для того чтобы это сделать, нужно вызвать от переменной *training_data* метод **head()**, который выводит первые 5 строк таблицы.

Для вызова метода объекта необходимо сначала написать *имя объекта*, затем поставить *точку*, затем уже написать *название метода*. Обратите внимание, что в конце обязательно ставить скобочки, потому что метод - это функция и в ней есть аргументы, просто в данном случае мы их не передаем, поэтому оставляем поле пустым

In [6]:
training_data.head()

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
0        830000        5    3.50           3490          21780   
1        385000        4    1.75           2360           7620   
2        610000        6    2.75           2040           8560   
3        550000        3    1.75           1940           8376   
4       1300000        3    2.75           3450           5350   

   Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
0                2.0            0                  0          3   
1                1.0            0                  0          4   
2                1.0            0                  2          4   
3                1.0            0                  0          4   
4                1.5            0                  3          4   

   Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
0                8                 3490                0           1996   
1                7                 1180             1180           1955   
2                7                 1100              940           1961   
3                8                 1290              650           1963   
4                9                 2590              860           1925   

   Год реновации   Широта  Долгота  
0              0  47.6707 -122.144  
1              0  47.5278 -122.345  
2              0  47.6160 -122.115  
3              0  47.5586 -122.173  
4              0  47.6389 -122.407

**Шаг 1.3. Посмотрим на размеры загруженной таблицы**, у которой мы видели только первые 5 строк.

Для этого вызываем поле **shape** у нашей переменной *training_data*. Поле вызывается также как метод, но в конце скобки не ставятся, так как для поля не предусмотрена передача аргументов.  

In [7]:
training_data.shape

(15129, 16)

*Что означает первое и второе число?*

Итак, таблица содержит 15129 строк (объектов) и 16 столбцов (признаков), включая выходной (целевой) признак.
15129 меньше 21613, так как мы пока загрузили только часть данных, которую будем использовать для обучения модели.

Таблицу проверили, теперь можно приступать к обработке данных.

# 2. Обработать данные перед обучением модели

**Шаг 2.1. Проверяем данные на наличие пропусков и типов переменных**

Начнем с проверки общей информации о данных.
Для того чтобы это сделать, нужно обратиться вызвать у переменной *training_data* метод **info()**.

Напомним, что в конце необходимо поставить скобочки.

In [8]:
training_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15129 entries, 0 to 15128
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Целевая.Цена         15129 non-null  int64  
 1   Спальни              15129 non-null  int64  
 2   Ванные               15129 non-null  float64
 3   Жилая площадь        15129 non-null  int64  
 4   Общая площадь        15129 non-null  int64  
 5   Количество этажей    15129 non-null  float64
 6   Вид на воду          15129 non-null  int64  
 7   Просмотрены ранее    15129 non-null  int64  
 8   Состояние            15129 non-null  int64  
 9   Оценка риелтора      15129 non-null  int64  
 10  Площадь без подвала  15129 non-null  int64  
 11  Площадь подвала      15129 non-null  int64  
 12  Год постройки        15129 non-null  int64  
 13  Год реновации        15129 non-null  int64  
 14  Широта               15129 non-null  float64
 15  Долгота              15129 non-null 

Анализируем результата выполнения команды:

* 15129 строк (entries)
* 16 столбцов (Data columns)

В данных присутствует всего два типа dtypes:
* int64 - целое число  (12 столбцов)
* float64 - дробное число (4 столбца)

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Так как эти цифры в каждой строчке совпадают с числом строк (15129), то в данных нет пропусков и можно двигаться дальше.

**Шаг 2.2. Работаем с целевой переменной**

*Какая переменная целевая?*

В данном случае по условию задачи мы должны прогнозировать стоимость, поэтому целевая переменная - это цена.

In [9]:
target_variable_name = 'Целевая.Цена'

Нам нужно выделить в отдельную переменную *training_values* столбец из нашей таблицы, который соответствует определенной выше целевой переменной. Для этого мы у таблицы *training_data* в квадратных скобках указываем имя нужного столбца. В нашем случае это имя записано в переменной *target_variable_name*.

In [10]:
training_values = training_data[target_variable_name]

Отделим входные переменные от выходной (целевой), чтобы можно было построить модель предсказания целевой переменной по входным.
Для это нужно у переменной *training_data* вызвать метод **drop()**. Результат мы записываем в новую переменную *training_points*. После выполнения запроса *training_points* будет содержать исходную таблицу без целевого столбца.

Обратите внимание, что в данном случае мы передаем два аргумента:
    1. target_variable_name - название столбца цены, который мы ранее записали в эту переменную и теперь хотим удалить из training_data
    2. axis=1 - означает, что мы удаляем столбец, а в случае axis=0 - означает, что мы удаляем строку

In [11]:
training_points = training_data.drop(target_variable_name, axis=1)

Можно посмотреть результаты этих действий, вызвав метод **head()** и поле **shape**, которыми мы пользовались ранее, но сейчас нужно вызывать их от новой переменной *training_points*.

In [12]:
training_points.head()

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0        5    3.50           3490          21780                2.0   
1        4    1.75           2360           7620                1.0   
2        6    2.75           2040           8560                1.0   
3        3    1.75           1940           8376                1.0   
4        3    2.75           3450           5350                1.5   

   Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0            0                  0          3                8   
1            0                  0          4                7   
2            0                  2          4                7   
3            0                  0          4                8   
4            0                  3          4                9   

   Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                 3490                0           1996              0   
1                 1180             1180           1955              0   
2                 1100              940           1961              0   
3                 1290              650           1963              0   
4                 2590              860           1925              0   

    Широта  Долгота  
0  47.6707 -122.144  
1  47.5278 -122.345  
2  47.6160 -122.115  
3  47.5586 -122.173  
4  47.6389 -122.407

In [13]:
training_points.shape

(15129, 15)

Видно, что столбца действительно нет, а количество строк не изменилось. Данные в 5 первых строках такие же, как были ранее.

##   3. Обучить модель на обучающей выборке

**Шаг 3.1. Выбираем метод, который будем использовать**

Проще всего начать с простых методов.

На выбор метода для построения модели влияет набор признаков, размер выборки, интуиция про то, какая связь между входными переменными и целевой. Но часто решение принимается исходя из того, какая модель сработала лучше.

Для корректной работы с методами построения моделей в python требуется загрузить специальную библиотеку
**sklearn**, программную библиотеку на языке python для для машинного обучения и анализа данных.

Импортируем модуль из этой библиотеки:
 * *linear_model* - тут находятся все линейные модели

In [14]:
from sklearn import linear_model

Чтобы создать модель линейной регресии, пишем имя модуля 'linear_model', затем точку, затем название модели.

In [15]:
linear_regression_model = linear_model.LinearRegression() # создаем модель

In [16]:
linear_regression_model # смотрим, что получилось

LinearRegression()

**Шаг 3.2. Обучить модель**

Теперь, когда мы создали прототип модели, можем её обучить на обучающей выборке.

Для этого вызываем метод **fit()** и передаем ему на вход два аргумента:
таблицу входных признаков и столбец значений целевой переменной - (training_points, training_values)

In [17]:
linear_regression_model.fit(training_points, training_values)

LinearRegression()

## Задание 3.5

Реализуйте настройку параметров модели не с помощью встроенной функции, а с помощью своей реализации.

Настроить параметры нужно либо с помощью **стохастического градиентного спуска** либо с помощью **mini-batch градиентного спуска**.
Нужно использовать регуляризацию **L1, L2 или Elastic Net**.

In [18]:
training_points # X

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0            5    3.50           3490          21780                2.0   
1            4    1.75           2360           7620                1.0   
2            6    2.75           2040           8560                1.0   
3            3    1.75           1940           8376                1.0   
4            3    2.75           3450           5350                1.5   
...        ...     ...            ...            ...                ...   
15124        4    4.25           5710          24663                2.0   
15125        3    1.50           1950           4473                1.0   
15126        3    2.00           3620          14250                2.0   
15127        4    2.50           2340           9748                1.0   
15128        3    1.50           1450           7316                1.0   

       Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0                0                  0          3                8   
1                0                  0          4                7   
2                0                  2          4                7   
3                0                  0          4                8   
4                0                  3          4                9   
...            ...                ...        ...              ...   
15124            0                  0          3               11   
15125            0                  0          4                6   
15126            0                  0          3                8   
15127            0                  1          3                8   
15128            0                  0          3                7   

       Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                     3490                0           1996              0   
1                     1180             1180           1955              0   
2                     1100              940           1961              0   
3                     1290              650           1963              0   
4                     2590              860           1925              0   
...                    ...              ...            ...            ...   
15124                 5710                0           2007              0   
15125                 1530              420           1914              0   
15126                 3220              400           2007              0   
15127                 1610              730           1981              0   
15128                 1450                0           1961              0   

        Широта  Долгота  
0      47.6707 -122.144  
1      47.5278 -122.345  
2      47.6160 -122.115  
3      47.5586 -122.173  
4      47.6389 -122.407  
...        ...      ...  
15124  47.3065 -122.158  
15125  47.7094 -122.228  
15126  47.5767 -122.225  
15127  47.3363 -122.331  
15128  47.7725 -122.349  

[15129 rows x 15 columns]

In [19]:
training_values # y

0         830000
1         385000
2         610000
3         550000
4        1300000
          ...   
15124     930000
15125     400000
15126    1580000
15127     339900
15128     430000
Name: Целевая.Цена, Length: 15129, dtype: int64

In [20]:
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error

In [21]:
def stochastic_gradient_descent(X, y, learning_rate, niterations):
  w = 0.001 * np.random.randn(X.shape[1])
  b = 0

  for _ in range(niterations):
    for i in range(X.shape[0]):
      y_predicted = predict(X[i], w, b)
      regularization_coefficient = 0.01

      # L2 = regularization_coefficient * (y_predicted - y[i])
      L2 = regularization_coefficient * np.sum(w ** 2)

      w -= learning_rate * L2 * X[i]
      b -= learning_rate * L2

  return w, b


def predict(X, w, b):
  y = 0

  for i in range(len(X)):
    y += w[i] * X[i]

  return y + b

# 4. Загрузить и предобработать данные для тестирования

**Шаг 4.1. Загрузим и проанализируем тестовые данные.**

Так как данные в формате xlsx (Excel), мы будем использовать специальную функцию
из библиотеки pandas для загрузки таких данных **read_excel**.

В функции передаем один атрибут: название файла, в котором находится таблица с данными.

In [22]:
test_data = pd.read_excel('predict_house_price_test_data.xlsx')

*Что важно посмотреть, после того, как мы загрузили данные?*

- проверить, что данные действительно загрузились
- посмотреть на данные, чтобы удостовериться, что они правильные: колонки имеют те же названия, что и в таблице и т.д.

Для того чтобы это сделать, нужно вызвать от переменной *test_data* метод **head()**, который выводит первые 5 строк таблицы.

Для вызова метода объекта необходимо сначала написать *имя объекта*, затем поставить *точку*, затем уже написать *название метода*. Обратите внимание, что в конце обязательно ставить скобочки, потому что метод - это функция и в ней есть аргументы, просто в данном случае мы их не передаем, поэтому оставляем поле пустым

In [23]:
test_data.head()

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
0        260000        3    1.00           1300          10139   
1        734500        4    2.75           3280           6845   
2        325000        1    1.00           1220          12426   
3       1990000        3    2.50           2880          13500   
4        315000        3    2.00           1300           3731   

   Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
0                1.0            0                  0          3   
1                2.0            0                  0          3   
2                1.0            0                  4          4   
3                1.0            0                  4          5   
4                1.0            0                  0          3   

   Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
0                7                 1300                0           1962   
1               10                 3280                0           2003   
2                6                 1220                0           1946   
3                8                 1520             1360           1950   
4                7                  900              400           1993   

   Год реновации   Широта  Долгота  
0           2007  47.3427 -122.087  
1              0  47.7042 -122.107  
2              0  47.4047 -122.331  
3              0  47.6281 -122.216  
4              0  47.5374 -122.270

Посмотрим на размеры загруженной таблицы, так как мы видели только 5 строк

Для этого вызываем поле **shape** у нашей переменной *test_data*. Поле вызывается также как метод, но в конце скобки не ставятся (!), так как для поля не предусмотрена передача аргументов.  

In [24]:
test_data.shape

(6484, 16)

*Что означает первое и второе число?*
Таблица содержит 6484 строк (объектов) и 16 столбцов (признаков), включая выходной (целевой) признак. Также как в учебных данных до обучения.

Таблицу проверили, теперь можно приступать к обработке данных
Действуем аналогично тому, как делали с данными для обучения

Проверим, есть ли в данных пропуски. Для того чтобы это сделать, нужно обратиться вызвать у переменной *test_data* метод **info()**.

In [25]:
test_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6484 entries, 0 to 6483
Data columns (total 16 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Целевая.Цена         6484 non-null   int64  
 1   Спальни              6484 non-null   int64  
 2   Ванные               6484 non-null   float64
 3   Жилая площадь        6484 non-null   int64  
 4   Общая площадь        6484 non-null   int64  
 5   Количество этажей    6484 non-null   float64
 6   Вид на воду          6484 non-null   int64  
 7   Просмотрены ранее    6484 non-null   int64  
 8   Состояние            6484 non-null   int64  
 9   Оценка риелтора      6484 non-null   int64  
 10  Площадь без подвала  6484 non-null   int64  
 11  Площадь подвала      6484 non-null   int64  
 12  Год постройки        6484 non-null   int64  
 13  Год реновации        6484 non-null   int64  
 14  Широта               6484 non-null   float64
 15  Долгота              6484 non-null   f

Цифры в каждой строчке обозначают количество заполненных (*non-null*) значений. Так как эти цифры в каждой строчке совпадают с числом строк (6484), то в данных нет пропусков.

**Шаг 4.2. Отделяем целевую переменную**

Нам нужно выделить в отдельную переменную *test_values* столбец из нашей таблицы, который соответствует определенной выше целевой переменной. Для этого мы у таблицы *test_data* в квадратных скобках указываем имя нужного столбца. В нашем случае это имя записано в переменной *target_variable_name*.

In [26]:
test_values = test_data[target_variable_name]

Отделим входные переменные от выходной (целевой), чтобы можно было построить модель предсказания целевой переменной по входным.
Для это нужно у переменной *test_data* вызвать метод **drop()**. Результат мы записываем в новую переменную *test_points*. После выполнения запроса *test_points* будет содержать исходную таблицу без целевого столбца.

Обратите внимание, что в данном случае мы передаем два аргумента:
    1. target_variable_name - название столбца цены, который мы ранее записали в эту переменную и теперь хотим удалить из training_data
    2. axis=1 - означает, что мы удаляем столбец, а в случае axis=0 - означает, что мы удаляем строку

In [27]:
test_points = test_data.drop(target_variable_name, axis=1)

И проверяем результат записанный в test_points

In [28]:
test_points.head()

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0        3    1.00           1300          10139                1.0   
1        4    2.75           3280           6845                2.0   
2        1    1.00           1220          12426                1.0   
3        3    2.50           2880          13500                1.0   
4        3    2.00           1300           3731                1.0   

   Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0            0                  0          3                7   
1            0                  0          3               10   
2            0                  4          4                6   
3            0                  4          5                8   
4            0                  0          3                7   

   Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                 1300                0           1962           2007   
1                 3280                0           2003              0   
2                 1220                0           1946              0   
3                 1520             1360           1950              0   
4                  900              400           1993              0   

    Широта  Долгота  
0  47.3427 -122.087  
1  47.7042 -122.107  
2  47.4047 -122.331  
3  47.6281 -122.216  
4  47.5374 -122.270

In [29]:
test_points.shape

(6484, 15)

Видно, что количество столбцов уменьшилось на один. Данные в 5 первых строках такие же, как были ранее.

# 5. Провалидировать модель на тестовой выборке

**Шаг 5.1. Сравнение моделей.**

Теперь мы готовы сравнить качество двух моделей! 😎

## Задание 3.6 Какая модель лучше?

Получим прогнозы целевой переменной на тестовых данных для модели линейной регрессии из sklearn и для своей реализации.



In [30]:
test_predictions_linear = linear_regression_model.predict(test_points)

In [31]:
### введите свой код
### код для получения предсказаний с помощью своей реализации алгоритма настройки параметров модели

scaler = StandardScaler()
test_points_for_sgd_model = scaler.fit_transform(test_points)

w, b = stochastic_gradient_descent(test_points_for_sgd_model, test_values, 0.001, 100)
test_predictions_linear_sgd = test_points_for_sgd_model.dot(w) + b

Качество регрессионных моделей оценим двумя способами:
1. Сравним визуально прогнозы с настоящими ценами (тестовые с предсказанием)
2. Сравним метрики качества MSE и MAE

Для корректного подсчета метрик качества модели в python требуется загрузить их из библиотеки **sklearn**.

Мы используем две метрики качества:
 * *mean_absolute_error* - средняя абсолютная ошибка $|y_i - \hat{y}_i|$
 * *mean_squared_error* - средняя квадратичная ошибка $(y_i - \hat{y}_i)^2$

In [32]:
from sklearn.metrics import mean_absolute_error, mean_squared_error

Подсчитаем ошибки для линейной модели.

Для этого вызовем методы **mean_absolute_error()** и **mean_squared_error()**. На вход им передается столбец настоящих значений *test_values* и столбец значений, предсказанных моделью линейной регрессии *test_predictions_linear*.

In [33]:
mean_absolute_error_linear_model = mean_absolute_error(test_values, test_predictions_linear)
mean_squared_error_linear_model = mean_squared_error(test_values, test_predictions_linear)

print("With sklearn model:\n")
print("MAE:", mean_absolute_error_linear_model)
print("MSE:", mean_squared_error_linear_model)

With sklearn model:

MAE: 126852.5125500516
MSE: 40756843765.1009


In [34]:
mean_absolute_error_linear_model_sgd = mean_absolute_error(test_values, test_predictions_linear_sgd)
mean_squared_error_linear_model_sgd = mean_squared_error(test_values, test_predictions_linear_sgd)

print("With custom SGD model:\n")
print("MAE:", mean_absolute_error_linear_model_sgd)
print("MSE:", mean_squared_error_linear_model_sgd)

With custom SGD model:

MAE: 543418.1010632125
MSE: 431552101908.4497


Подсчитайте ошибки для своей реализации модели.

Теперь напечатаем полученные ошибки. Обычно смотрят на корень из среднеквадратичной ошибки, RMSE. Чтобы извлечь корень нам понадобится библиотека **Numpy**. С помощью неё можно быстро производить вычисления сразу над массивами чисел.

In [55]:
print("RMSE for sklearn linear regression model:", np.sqrt(mean_squared_error_linear_model))
print("RMSE for custom SGD linear regression model:", np.sqrt(mean_squared_error_linear_model_sgd))

RMSE for sklearn linear regression model: 201883.24290317137
RMSE for custom SGD linear regression model: 343200.99996416585


## 6. Выявление важных признаков

Признаки, используемые для обучения модели, оказывают большое влияние на качество результатов. Неинформативные или слабо информативные признаки могут снизить качество модели, использование избыточных признаков в модели замедляет скорость ее работы.

Для того, чтобы отобрать признаки нам следует ответить на следующие вопросы:

    1. Все ли признаки в наших данных заполненны разумными значениями?
    2. Какие признаки будут больше всего влиять на значение целевой переменной?
    3. Какие дополнительные признаки имело бы смысл добавить в список входных?

**6.1. Разглядывание значений признаков**

Для того чтобы ответить на первый вопрос, нам необходимо получить некоторую нашу экспертную оценку данных. В наших данных нет пропусков, однако иногда пропуски предварительно заполняют другими значениями (0, 999 и другие) и их можно увидеть, только внимательно посмотрев на данные.

Для того, чтобы это сделать, нужно вызвать у переменной *training_points* метод **head(10)**, который выводит первые 10 строк таблицы.

In [56]:
training_points.head(10)

Спальни  Ванные  Жилая площадь  Общая площадь  Количество этажей  \
0        5    3.50           3490          21780                2.0   
1        4    1.75           2360           7620                1.0   
2        6    2.75           2040           8560                1.0   
3        3    1.75           1940           8376                1.0   
4        3    2.75           3450           5350                1.5   
5        3    2.25           2300           9914                2.0   
6        3    2.50           2770           8820                1.0   
7        2    1.75           1650           7500                1.0   
8        1    1.00            580           1799                1.0   
9        2    1.00            900           3400                1.0   

   Вид на воду  Просмотрены ранее  Состояние  Оценка риелтора  \
0            0                  0          3                8   
1            0                  0          4                7   
2            0                  2          4                7   
3            0                  0          4                8   
4            0                  3          4                9   
5            0                  0          4                8   
6            0                  0          3                7   
7            0                  0          4                7   
8            0                  0          3                7   
9            0                  0          5                6   

   Площадь без подвала  Площадь подвала  Год постройки  Год реновации  \
0                 3490                0           1996              0   
1                 1180             1180           1955              0   
2                 1100              940           1961              0   
3                 1290              650           1963              0   
4                 2590              860           1925              0   
5                 2300                0           1980              0   
6                 1900              870           1980           2004   
7                 1000              650           1959              0   
8                  580                0           1908           2005   
9                  900                0           1905              0   

    Широта  Долгота  
0  47.6707 -122.144  
1  47.5278 -122.345  
2  47.6160 -122.115  
3  47.5586 -122.173  
4  47.6389 -122.407  
5  47.5677 -122.086  
6  47.3685 -122.048  
7  47.6871 -122.207  
8  47.6829 -122.375  
9  47.5269 -122.314

Рассмотрим столбец "Год реновации". Он содержит много нулей и мало реальных данных о годах, что как раз свидетельствует о том, что в этих данных изначально были пропуски, а к нам они попали уже заполненные нулями.

Мы можем посмотреть количество уникальных значений и сколько раз эти значения встречаются в этом столбце. Для этого вызываем метод **value_counts()** у нашего столбца *training_points['Год реновации']*

In [57]:
training_points['Год реновации'].value_counts()

0       14490
2014       63
2013       31
2000       28
2003       24
        ...  
1948        1
1950        1
1953        1
1934        1
1955        1
Name: Год реновации, Length: 67, dtype: int64

Как видно, нулей существенно больше (14490), чем любых других значений (639), поэтому можем сделать вывод, что этот признак не информативен и может негативно повлиять на качество модели.

**6.2. Какие признаки самые важные**

Ручной перебор признаков невозможен, если их тысячи, кроме того, часто даже эксперты не могут сказать, использование каких признаков в модели даст прирост качества.
Кроме ручного перебора всех признаков и их визуального анализа, можно использовать оценку значимости признаков с помощью обученных моделей.

## Задание 3.7

In [58]:
### обучаем модель на предобработанных данных

scaler = StandardScaler()
test_points_fit_and_transformed = scaler.fit_transform(test_points)
linear_regression_model.fit(test_points_fit_and_transformed, test_values)

LinearRegression()

Представьте полученные результаты в более удобном табличном виде. Создайте пустую таблицу *feature_importance* с двумя колонками "Название признака", "Важность признака". Для этого воспользуемся функцией **DataFrame()** из библиотеки pandas и передадим ей в качестве аргумента   
*columns = ["Название признака", "Важность признака"]*

In [59]:
feature_importance = pd.DataFrame(columns = ['Название признака', 'Важность признака'])

Заполните соответствующие колонки значениями, которые мы получили выше.

In [60]:
feature_importance['Название признака'] = training_points.columns
feature_importance['Важность признака'] = linear_regression_model.coef_

feature_importance

Название признака  Важность признака
0               Спальни      -31097.717444
1                Ванные       30497.441929
2         Жилая площадь       81838.845031
3         Общая площадь       -4272.116365
4     Количество этажей        4580.865341
5           Вид на воду       49260.324104
6     Просмотрены ранее       43881.689515
7             Состояние       16928.790580
8       Оценка риелтора      129109.539695
9   Площадь без подвала       74919.909157
10      Площадь подвала       31118.359211
11        Год постройки      -74355.867736
12        Год реновации        3507.520922
13               Широта       76236.339799
14              Долгота      -12538.241338

Мы можем отсортировать значения по важности, вызвав у *feature_importance* метод **sort_values()**.

Обратите внимание, что в данном случае мы передаем два аргумента:

1. by='Важность признака' - название столбца, по которому мы сортируем
2. ascending=False - означает, что мы сортируем по убыванию, а в случае ascending=True сортировка будет по возрастанию

In [61]:
feature_importance.sort_values(by = 'Важность признака', ascending = False)

Название признака  Важность признака
8       Оценка риелтора      129109.539695
2         Жилая площадь       81838.845031
13               Широта       76236.339799
9   Площадь без подвала       74919.909157
5           Вид на воду       49260.324104
6     Просмотрены ранее       43881.689515
10      Площадь подвала       31118.359211
1                Ванные       30497.441929
7             Состояние       16928.790580
4     Количество этажей        4580.865341
12        Год реновации        3507.520922
3         Общая площадь       -4272.116365
14              Долгота      -12538.241338
0               Спальни      -31097.717444
11        Год постройки      -74355.867736

In [62]:
feature_importance.reindex(feature_importance['Важность признака'].abs().sort_values(ascending = False).index)

Название признака  Важность признака
8       Оценка риелтора      129109.539695
2         Жилая площадь       81838.845031
13               Широта       76236.339799
9   Площадь без подвала       74919.909157
11        Год постройки      -74355.867736
5           Вид на воду       49260.324104
6     Просмотрены ранее       43881.689515
10      Площадь подвала       31118.359211
0               Спальни      -31097.717444
1                Ванные       30497.441929
7             Состояние       16928.790580
14              Долгота      -12538.241338
4     Количество этажей        4580.865341
3         Общая площадь       -4272.116365
12        Год реновации        3507.520922

## Задание 3.8.

Напомню исходную постановку задачи. Мы говорили о том, что оценка цены на недвижимость может быть нужна для выявления аномально низких цен на объект недвижимости. Давайте попробуем найти такие в датасете.

In [63]:
linear_regression_model.fit(training_points, training_values)
predictions = linear_regression_model.predict(training_points)

In [64]:
mae = mean_absolute_error(training_values, predictions)
mse = mean_squared_error(training_values, predictions)

print(f"MAE: {mae} , MSE: {mse} , RMSE: {np.sqrt(mse)}")

MAE: 126760.36255980784 , MSE: 41513370049.359474 , RMSE: 203748.30072753853


In [65]:
data = training_data.copy()
data["Предсказанная.Цена"] = predictions
data

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
0            830000        5    3.50           3490          21780   
1            385000        4    1.75           2360           7620   
2            610000        6    2.75           2040           8560   
3            550000        3    1.75           1940           8376   
4           1300000        3    2.75           3450           5350   
...             ...      ...     ...            ...            ...   
15124        930000        4    4.25           5710          24663   
15125        400000        3    1.50           1950           4473   
15126       1580000        3    2.00           3620          14250   
15127        339900        4    2.50           2340           9748   
15128        430000        3    1.50           1450           7316   

       Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
0                    2.0            0                  0          3   
1                    1.0            0                  0          4   
2                    1.0            0                  2          4   
3                    1.0            0                  0          4   
4                    1.5            0                  3          4   
...                  ...          ...                ...        ...   
15124                2.0            0                  0          3   
15125                1.0            0                  0          4   
15126                2.0            0                  0          3   
15127                1.0            0                  1          3   
15128                1.0            0                  0          3   

       Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
0                    8                 3490                0           1996   
1                    7                 1180             1180           1955   
2                    7                 1100              940           1961   
3                    8                 1290              650           1963   
4                    9                 2590              860           1925   
...                ...                  ...              ...            ...   
15124               11                 5710                0           2007   
15125                6                 1530              420           1914   
15126                8                 3220              400           2007   
15127                8                 1610              730           1981   
15128                7                 1450                0           1961   

       Год реновации   Широта  Долгота  Предсказанная.Цена  
0                  0  47.6707 -122.144        8.262192e+05  
1                  0  47.5278 -122.345        4.913247e+05  
2                  0  47.6160 -122.115        5.253505e+05  
3                  0  47.5586 -122.173        5.456459e+05  
4                  0  47.6389 -122.407        1.284459e+06  
...              ...      ...      ...                 ...  
15124              0  47.3065 -122.158        1.391401e+06  
15125              0  47.7094 -122.228        5.600209e+05  
15126              0  47.5767 -122.225        7.651392e+05  
15127              0  47.3363 -122.331        4.824668e+05  
15128              0  47.7725 -122.349        4.824416e+05  

[15129 rows x 17 columns]

In [67]:
target = data[data["Целевая.Цена"] <= data["Предсказанная.Цена"] - 2 * mae].sort_values(by = "Целевая.Цена", ascending = False)
target

Целевая.Цена  Спальни  Ванные  Жилая площадь  Общая площадь  \
5700        1680000        3    2.50           3490           8343   
8005        1640000        3    3.50           4660          21164   
9631        1610000        4    2.75           5470          18200   
3497        1580000        4    2.50           4570          74487   
10543       1490000        5    6.00           6880         279968   
...             ...      ...     ...            ...            ...   
9496         170000        2    1.00           1500          18540   
11563        169317        2    1.00            790           4000   
12715        160000        2    1.00           1140          23030   
17           154000        2    1.00           1040          20524   
932          134000        2    1.50            980           5000   

       Количество этажей  Вид на воду  Просмотрены ранее  Состояние  \
5700                 2.0            1                  4          4   
8005                 2.0            1                  4          3   
9631                 2.0            1                  4          3   
3497                 2.0            0                  4          5   
10543                2.0            0                  3          3   
...                  ...          ...                ...        ...   
9496                 1.0            0                  0          3   
11563                1.0            0                  2          3   
12715                1.0            0                  0          3   
17                   1.0            0                  3          3   
932                  2.0            0                  0          3   

       Оценка риелтора  Площадь без подвала  Площадь подвала  Год постройки  \
5700                 9                 2150             1340           1939   
8005                12                 4660                0           1975   
9631                11                 3730             1740           1992   
3497                12                 4570                0           1948   
10543               12                 4070             2810           2007   
...                ...                  ...              ...            ...   
9496                 8                 1500                0           1950   
11563                7                  790                0           1908   
12715                8                 1140                0           1980   
17                   6                 1040                0           1949   
932                  7                  980                0           1922   

       Год реновации   Широта  Долгота  Предсказанная.Цена  
5700            1991  47.7265 -122.281        1.949411e+06  
8005            1990  47.4418 -122.354        2.304172e+06  
9631               0  47.5316 -122.263        2.170063e+06  
3497            1985  47.7282 -122.372        1.911047e+06  
10543              0  47.4624 -121.779        1.821101e+06  
...              ...      ...      ...                 ...  
9496               0  47.4727 -122.281        4.519697e+05  
11563              0  47.6248 -122.301        5.057037e+05  
12715              0  47.7637 -122.266        4.724389e+05  
17              1989  47.6981 -121.875        4.480426e+05  
932             2003  47.7076 -121.359        4.186541e+05  

[760 rows x 17 columns]